### ISSUE TITLE
avatarDownload() - TCP Connection not closed

### ISSUE THREAD

**Issue by Newcomer1989** [planetteamspeak/ts3phpframework#14]
I wrote a bot, which downloads the avatars by getting this from the filelist (cid=0).
`$ts3->channelFileList($cid="0", $cpw="", $path="/");`

I took the passage from the avatarDownload() function and only modified it a bit, because the filelist give me already the file name that is needed for the download.

```
$avatar = $ts3->transferInitDownload($clientftfid="5",$cid="0",$name=$fullfilename,$cpw="", seekpos=0);
$transfer = TeamSpeak3::factory("filetransfer://" . $avatar["host"] . ":" . $avatar["port"]);
$tsfile = $transfer->download($avatar["ftkey"], $avatar["size"]);
$avatarfilepath	= substr(__DIR__,0,-4).'avatars/'.$uuidasbase16.'.png';
```

This works great. The avatar gets downloaded like wished.

For each download of an avatar there will be opened a TCP connection to the ts3 server. That's ok and correct. However I think this connection will not be closed by the framework.
I see there stands for each downloaded avatar a connection on status "CLOSE_WAIT".
The problem is that my script runs as bot, so the connection will not be closed until the bot runs.

> 
tcp        0      0 xxx.xxx.xxx.xxx:38539    xxx.xxx.xxx.xxx:30033   CLOSE_WAIT  19931/php

Did I something wrong in my code or the handling of the TCP connections inside the framework are the problem?

**Comment by Newcomer1989** [comment]
any feedback?

**Comment by ronindesign** [comment]
I see the sock init in `TeamSpeak3\Transport\TeamSpeak3_Transport_TCP` for file transfer transport:

```php
$this->stream = @stream_socket_client($address, $errno, $errstr, $timeout);
...
@stream_set_timeout($this->stream, $timeout);
```

The latter line implies the stream should timeout after `$timeout` which seems to be set to 10 seconds in `TeamSpeak3\Transport\TeamSpeak3_Transport_Abstract`:
```php
 if(!array_key_exists("timeout", $config))
    {
      $config["timeout"] = 10;
    }
```

I did see a note regarding stream / sockets that socket timeout might need to be set via `socket_set_option`

Otherwise, I didn't see any explicit calls to close streams, so assuming streams are meant to be dereferenced and allowed to timeout after use. However, if the timeout option is not working as expected, then that would definitely attribute to what you are describing.

**Comment by ronindesign** [comment]
I've done some initial testing on this now.

Looks like multiple inits via `$transfer = TeamSpeak3::factory(...);` causes dangling references even when trying to destroy the file transfer's adapter manually calling `unset($transfer);`.

Garbage collection does not seem to destruct the dereferenced adapter instance either, most likely because of the nature how the php script / bot is being ran.

To show proper, expected behavior, you can see that calling the destruct method directly (not preferable) `$transfer->__destruct()` does call the transport's disconnect() method and correctly closes out the connection.

For now, a breakfix might be to call the destructor manually, or reuse the same instance of `$transfer` while updating the `$avatar` to point at the next avatar to download. I've not tested if the latter will work, but ideally it will leave you with only one dangling `$transfer` instance, however with the timeout set, you may not be able to re-use this instance after the connection has dropped.

Another option outside the scope of the TeamSpeak3 library might be to call the garbage collector directly via: `gc_collect_cycles();` I've tested this and it does clean up any dereferenced connctions via their __destruct() method. However, you may need to examine the impact on the rest of your script or otherwise.

References:
http://stackoverflow.com/questions/8798443/best-way-to-destroy-php-object
http://www.stoimen.com/blog/2011/11/14/php-dont-call-the-destructor-explicitly/
http://php.net/manual/en/function.gc-collect-cycles.php

**Comment by Newcomer1989** [comment]
I made a quick test with the __destruct function. It works well. The connection got closed correct.

In my case it would not the best idea to halt the connection. Also Teamspeak changes the TTL with server version 3.0.13.

> === Server Release 3.0.13  1 august 2016
>  * file transfer now only allows an idle time of 2 seconds after connect

Perhaps it is reasonable to add the destruct also inside the framework for each filetransfer per default.

**Comment by ronindesign** [comment]
Typically calling `__destruct` directly is not ideal, using `gc_collect_cycles();` each interval after iterating over your set of objects (e.g. clients) would be more appropriate.

It does appear that the nature of garbage collection causes unintended results when external effects are present (e.g. socket connection) along with cyclic references.

For example, simply trying to unset, garbage collect, and thus destroy a single adapter instance of `TeamSpeak3_Adapter_ServerQuery`:
```php
$ts3 = TeamSpeak3::factory("serverquery:// ... ");
unset($ts3);
echo gc_collect_cycles();
```
Results in an output of `17` garbage cycles being collected. This means unset is working as expected, however because there are cyclic references, objects fail to be immediately garbage collected (because reference(s) still exist). I believe this is because collection for cyclic references does not occur until "the root buffer is full":
> To avoid having to call the checking of garbage cycles with every possible decrease of a refcount, the algorithm instead puts all possible roots (zvals) in the "root buffer" (marking them "purple"). It also makes sure that each possible garbage root ends up in the buffer only once. Only when the root buffer is full does the collection mechanism start for all the different zvals inside. See step A in the figure above.

Ref: http://php.net/manual/en/features.gc.collecting-cycles.php

However, in the end, it's still unclear which scope is responsible for handling full destruction of objects in a persistent environment.

My view is that the scope responsible for managing this is either the one in which an instance is created or its direct parent.
In this case, the persistent script's block that initializes any number of `$transfer` objects should clean up after. For example from previous comment:
```php
foreach($clients as $client) {
  $avatar = $ts3->transferInitDownload($client->clid,$cid="0",$name=$fullfilename,$cpw="", seekpos=0);
  $transfer = TeamSpeak3::factory("filetransfer://" . $avatar["host"] . ":" . $avatar["port"]);
  $tsfile = $transfer->download($avatar["ftkey"], $avatar["size"]);
  $avatarfilepath = substr(__DIR__,0,-4).'avatars/'.$uuidasbase16.'.png';
}
gc_collect_cycles();
```
Otherwise, it's not clear where in the framework this logic should go.
Should an instance of an adapter really be responsible for garbage collecting itself? Doesn't seem like the appropriate separation of concerns.
On the other hand, if the framework is instantiating instances such as in `TeamSpeak3_Node_Client->avatarDownload()`, maybe it should handle forcing garbage collection, for example:
```php
public function avatarDownload()
{
  if($this["client_flag_avatar"] == 0) return;

  $download = $this->getParent()->transferInitDownload(rand(0x0000, 0xFFFF), 0, $this->avatarGetName());
  $transfer = TeamSpeak3::factory("filetransfer://" . $download["host"] . ":" . $download["port"]);

  $avatar = $transfer->download($download["ftkey"], $download["size"]);
  unset($transfer);
  gc_collect_cycles();

  return $avatar;
}
```

This has serious implications though because it effects far outside the scope of the instance and again violates appropriate separation of concerns.

Additionally, it could have unintended or undesirable effects the the global scope of the running script and may cause serious difficulty when trying to debug a different, seemingly unrelated part of the script.

Most likely, the existing cyclic references that are created may very well be unintended in the first place.
I will open a ticket regarding this for further review, testing, and possibly development if a patch would fix this and allow for expected, appropriate garbage collection behavior.

**Comment by svenpaulsen** [comment]
Fixed with pull request #16.
